src/topics/resolver/telegram-resolver.ts

```diff
@@ -0,0 +1,160 @@
+import { BigNumberish } from "ethers";
+import { Api, TelegramClient } from "telegram";
+import { LogLevel } from "telegram/extensions/Logger";
+import { StringSession } from "telegram/sessions";
+import { IResolver } from "./resolver";
+import {
+  resolveAccount,
+  withConcurrency,
+  handleResolvingErrors,
+} from "./utils";
+import { FetchedData } from "topics/group";
+
+export class TelegramResolver implements IResolver {
+  private _apiHash: string;
+  private _apiId: string;
+  private _botSession: string | undefined;
+  private _botToken: string;
+  private _client: TelegramClient;
+
+  constructor(
+    apiHash = process.env.TELEGRAM_API_HASH || "",
+    apiId = process.env.TELEGRAM_API_ID || "",
+    botToken = process.env.TELEGRAM_BOT_TOKEN || "",
+    botSession = process.env.TELEGRAM_BOT_SESSION
+  ) {
+    this._apiHash = apiHash;
+    this._apiId = apiId;
+    this._botSession = botSession;
+    this._botToken = botToken;
+  }
+
+  public resolve = async (
+    accounts: FetchedData
+  ): Promise<[FetchedData, FetchedData]> => {
+    const alreadyUpdatedAccounts: FetchedData = {};
+    const alreadyResolvedAccounts: FetchedData = {};
+
+    // extract usernames already resolved
+    const unresolvedAccounts = Object.entries(accounts).filter(
+      ([account, value]) => {
+        if (account.split(":").length === 3) {
+          const id = account.split(":")[2];
+          alreadyResolvedAccounts[resolveAccount("1003", id)] = value;
+          alreadyUpdatedAccounts[account] = value;
+        }
+        return account.split(":").length !== 3;
+      }
+    );
+
+    await this._initTelegramClient();
```

Comment:
The initiation of the client should be done in the constructor

Reply:
as well as the connection

Reply:
I've moved the call to conect() in _initTelegramClient.

About moving it in the constructor, we cannot call async functions in the constructor because it complains that the constructor cannot be async. While I can probably work around this, I think deferring the connection to as late as possible is not a bad idea. Otherwise I think we risk having a connection open even if the resolver is not used, because the class is instantiated in ResolverFactory all the time.

Reply:
I've moved the init in the constructor except for the calls to sign in and connect. Those we must not do in the constructor because the constructor is called when the ResolverFactory object is created. This means we would have Telegram sign in and connect even when it's not needed.